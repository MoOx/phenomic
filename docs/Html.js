import React from "react";
import Head from "react-helmet";
import { StyleSheet } from "react-primitives";

export type HtmlPropsType = {
  body: React$Element<*>,
  state?: React$Element<*>,
  script: React$Element<*>
};

const Html = (props: HtmlPropsType) => {
  const helmet = Head.renderStatic();
  return (
    <html {...helmet.htmlAttributes.toComponent()}>
      <head>
        {helmet.base.toComponent()}
        {helmet.title.toComponent()}
        {helmet.meta.toComponent()}
        <link rel="stylesheet" href="/styles.css" />
        {StyleSheet.getStyleSheets().map(({ id, textContent }) =>
          <style key={id} id={id}>{textContent}</style>
        )}
        {helmet.link.toComponent()}
        {helmet.style.toComponent()}
        {helmet.script.toComponent()}
        {helmet.noscript.toComponent()}
      </head>
      <body {...helmet.bodyAttributes.toComponent()}>
        {props.body}
        {props.state}
        {props.script}
      </body>
    </html>
  );
};

export default Html;
