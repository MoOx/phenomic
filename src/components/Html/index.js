// @flow

import React from "react"
import { renderToString } from "react-dom/server"
import Helmet from "react-helmet"

type Props = {
  css: Array<string>,
  js: Array<string>,
  renderBody: () => React$Element<any>,
  renderScript: () => React$Element<any>,
}

const Html = (props: Props) => {

  // Glamor integration
  // https://github.com/threepointone/glamor/blob/master/docs/server.md
  let glamorRenderStatic
  try {
    // $FlowFixMe just ignore glamor as we don't have it as a dep
    glamorRenderStatic = require("glamor/server").renderStatic
  }
  catch (e) {
    // skip glamor if not working
  }

  // Aprodite
  // https://github.com/Khan/aphrodite#server-side-rendering
  let aproditeRenderStatic
  try {
    // $FlowFixMe just ignore aprodite as we don't have it as a dep
    aproditeRenderStatic = require("aphrodite").StyleSheetServer.renderStatic
  }
  catch (e) {
    // skip aprodite if not working
  }

  // render body
  let body
  if (glamorRenderStatic) {
    const glamorResult = glamorRenderStatic(() => props.renderBody())

    renderToString(
      <Helmet
        style={ [
          { "cssText": glamorResult.css },
        ] }
        script={ [
          { "innerHTML": `window._glamor = ${
            JSON.stringify(glamorResult.ids)
          }` },
        ] }
      />
    )
    body = glamorResult.html
  }
  else if (aproditeRenderStatic) {
    const aproditeResult = aproditeRenderStatic(() => props.renderBody())

    renderToString(
      <Helmet
        style={ [
          {
            "cssText": aproditeResult.css.content,
            "data-aphrodite": undefined,
          },
        ] }
        script={ [
          { "innerHTML": `window._aphrodite = ${
            JSON.stringify(aproditeResult.css.renderedClassNames)
          };` },
        ] }
      />
    )
    body = aproditeResult.html
  }

  body = body || props.renderBody()

  // rewind html metas
  const head = Helmet.rewind()

  // <!doctype html> is automatically prepended
  return (
    <html
      lang="en"
      { ...head.htmlAttributes.toComponent() }
    >
      <head>
        { head.base.toComponent() }
        { head.title.toComponent() }
        <meta charSet="utf-8" />
        <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        { head.meta.toComponent() }
        { head.style.toComponent() }
        { head.link.toComponent() }
        {
          props.css.map((file, i) => (
            <link key={ "phenomic.css." + i } rel="stylesheet" href={ file } />
          ))
        }
        { head.script.toComponent() }
      </head>
      <body>
        <div id="phenomic" dangerouslySetInnerHTML={{ __html: body }} />
        { props.renderScript() }
        {
          props.js.map((file, i) => (
            <script key={ "phenomic.js." + i } src={ file } />
          ))
        }
      </body>
    </html>
  )
}

export default Html
