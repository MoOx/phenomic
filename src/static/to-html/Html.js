// @flow
import React from "react"

type Props = {
  htmlProps: Object,
  head: string,
  body: string,
  script: string,
  children: any,
}

const Html = (props: Props) => (
  <html { ...props.htmlProps }>
    <head
      dangerouslySetInnerHTML={{
        __html: props.head,
      }}
    />
    <body>
      <div
        id="phenomic"
        dangerouslySetInnerHTML={{
          __html: props.body,
        }}
      />
      <script
        dangerouslySetInnerHTML={{
          __html: props.script,
        }}
      />
      { props.children }
    </body>
  </html>
)

export default Html
