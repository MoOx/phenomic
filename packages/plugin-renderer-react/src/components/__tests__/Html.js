/* eslint-disable react/no-multi-comp */

import React from "react";
import { renderJSX } from "jsx-test-helpers";

import Html from "../HTML";

const App = () => <div>{"App"}</div>;
const render = App => {
  const Main = App;
  const State = () => <script type="text/json">{'{st:"ate"}'}</script>;
  const Script = () => <script src="sc/ript" />;
  return {
    Main,
    State,
    Script,
    Body: () =>
      <body>
        <Main />
        <State />
        <Script />
      </body>
  };
};

test("Html works", () => {
  expect(renderJSX(<Html App={App} render={render} />)).toMatchSnapshot();
});
