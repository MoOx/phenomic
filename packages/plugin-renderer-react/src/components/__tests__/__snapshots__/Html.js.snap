// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Html works 1`] = `
"<html>
  <head>
    <meta charSet=\\"utf-8\\" />
    <meta
      content=\\"width=device-width, initial-scale=1\\"
      name=\\"viewport\\"
    />
  </head>
  <Body />
</html>"
`;
